### ISSUE TITLE
Dimension reduction and differential testing with diffcyt

### ISSUE THREAD

**Issue by sudu87** [HelenaLC/CATALYST#121]
Hello Helena,

This is the continuation of previous solved issue.
https://github.com/HelenaLC/CATALYST/issues/119

I am able to use `sce` object for dimension reduction and look at plots in case of `TSNE` and `UMAP`.
### Dimensionality reduction with 4000 cells per sample. NOTE: I have a lot more than this. However, my decently powerful Mac works overnight if I use all cells.
```r
set.seed(123)
sce <- runDR(sce, dr = "UMAP",cells = 4000, features = "type")
```
### Checking reduced dimensions
```r
reducedDimNames(sce)
```
      ##  [1] "UMAP"

```r
head(reducedDim(sce, "UMAP"))
```

      ##       [,1] [,2]
      ##  [1,]   NA   NA
      ##  [2,]   NA   NA
      ##  [3,]   NA   NA
      ##  [4,]   NA   NA
      ##  [5,]   NA   NA
      ##  [6,]   NA   NA
```r
tail(reducedDim(sce, "UMAP"))
```

      ##                [,1]      [,2]
      ##  [578650,] 1.430724 0.2447781
      ##  [578651,]       NA        NA
      ##  [578652,]       NA        NA
      ##  [578653,]       NA        NA
      ##  [578654,]       NA        NA
      ##  [578655,]       NA        NA

### Differential testing with diffcyt
```r
# create design & constrast matrix
design <- createDesignMatrix(ei(sce), cols_design = "condition")
contrast <- createContrast(c(0, 1))

# test for
# - differential abundance (DA) of clusters
# - differential states (DS) within clusters
res_DA <- diffcyt(sce, clustering_to_use = "meta10",
    analysis_type = "DA", method_DA = "diffcyt-DA-edgeR",
    design = design, contrast = contrast, verbose = FALSE)
res_DS <- diffcyt(sce, clustering_to_use = "meta10",
    analysis_type = "DS", method_DS = "diffcyt-DS-limma",
    design = design, contrast = contrast, verbose = FALSE)
```

`Error in glmFit.default(y = y$counts, design = design, dispersion = dispersion, : NA dispersions not allowed`

I am guessing this is due to all the NAs. If yes then why does this happen and how to get rid of NAs here ?

**Comment by HelenaLC** [comment]
1. Differential testing is completely independent of the dimensionality reduction. The NAs in `reducedDims` are due to the sampling of 4k cells per sample, but will not affect any downstream analysis.
2. By default, `diffcyt` will try and test the `state_markers(sce)`, unless argument `markers_to_test` is specified. Maybe worth checking that `state_markers(sce)` returns something? And that `rowData(sce)` has a column `marker_class` specified.
3. `diffcyt` is not `CATALYST`, and my ability to help here - beyond thinking out loud - is limited. Please post `diffcyt`-related issues here: https://github.com/lmweber/diffcyt/issues so @lmweber can help.

**Comment by sudu87** [comment]
Hello Helena, thanks for replying. I don't state markers as such because my experiment design has 16 surface markers (type) and 7 conditions. Is state mention necessary ? wouldn't it depend on the experiment design? 

`rowData(sce)`

      ##   DataFrame with 16 rows and 4 columns
      ##  channel_name marker_name marker_class used_for_clustering
      ##                 <character> <character> <factor>         <logical>
      ##   Siglec-8      Dy164Di    Siglec-8         type                TRUE
      ##   CD206         Er168Di       CD206         type                TRUE
      ##   CD169         Er170Di       CD169         type                TRUE
      ##   CD11b         Eu151Di       CD11b         type                TRUE
      ##   HLA-DR        Gd156Di      HLA-DR         type                TRUE
      ##   ...               ...         ...          ...                 ...
      ##   PD-L1         Sm154Di       PD-L1         type                TRUE
      ##   CD163         Tm169Di       CD163         type                TRUE
      ##   CD86          Yb171Di        CD86         type                TRUE
      ##   CD81          Yb173Di        CD81         type                TRUE
      ##   CD88          Yb174Di        CD88         type                TRUE

I will link this issue with the `diffcyt` page. 

I will also try to the argument `markers_to_test `

**Comment by HelenaLC** [comment]
Well, it is necessary in that `diffcyt` will test the subset of markers with `marker_class` "state", and `CATALYST`'s `cluster()` will default to using the "type" markers (in both cases, unless specified otherwise). So, there will have a strong impact on subpopulation definition and differential analysis. 
Yes, I think trying out `markers_to_test` might solve it, otherwise I suggest posting this as a `diffcyt` issue. I am a bit surprised though... If you have NO state markers defined, there should definitely be a warning for this.

**Comment by sudu87** [comment]
I am just wondering what `state `should be in an experiment like this. I have 2 types of cells monocytes (2 samples) and macrophages (5 samples). The monocytes have been differentiated into macrophages so may be " differentiated " and "no-differentiation"?

**Comment by HelenaLC** [comment]
I'd say this depends on your biological question- having specified all 15 markers as "type" I'd first ask: Do you really use all 16 markers to distinguish between monocytes & macrophages, and intermediate populations of these? I can imaging only a subset of these is used to tell them apart. I.e., you shouldn't need 10+ markers to tell apart 2-3 subtypes... What is the question you  want to ask- are you interested in how treatment affects expression, or how treatment affect the presence/absence of intermediate subpopulations or differentiation etc.? 
In summary: Yes, you are correct, all this depends on your design and question. My point above was simply that you "need" to define this if you don't use `markers_to_test` (in `diffcyt`) and `features` (in `cluster`), because most functions will default to using certain markers.

**Comment by sudu87** [comment]
Thanks for your critique Helena. This is definitely an important point. 
The goal is not to differentiate them rather its more to look at the populations coming up upon treatment of macrophages.
Control cells - Monocytes and Monocytes with DMSO treatment (2 samples)
Actual test cells - Macrophages (5 sanples).
Conditions for Macrophages - No treatment, LPS, PepG, LPS + PepG and Steroidal drug (GCM). 
I am interested in how each of these treatment influence heterogeneity of Macrophages.

**Comment by HelenaLC** [comment]
Ok. My immunological knowledge is (really) limited, so I’m just thinking out loud here to maybe help you think through some of this:
In one of my collaborator’s data, we’re interested in monocyte subsets, so classical non-classical intermediate. In this example, we cluster using canonical markers (cd14, cd16 and a couple others). I suggest a similar procedure here: Ask urself which markers you’d use to characterize the different subpopulations, and use only these. One issue with using all markers is that you introduce “noise”, ie differences / clusters that may not be biologically meaningful, while you miss the subsets that are. 
Then, I’d pay careful attention to the output of plotAbundances- in the monocyte project I mentioned, we saw that most “heterogeneity” was really only present in 1 sample. So, make sure the “new” subtypes you see are represented across samples, not just artifacts from some batch effect / inter-sample variability. 
Finally, if you’re interested in how the heterogeneity changes, I’d focus on getting a good (!!) clustering with meaningful subtypes first, using a subset of markers that will enable you to really annotate these subtypes (E.g. can you give them a meaningful annotation beyond cluster 1,2,3...?). Then, focus on visualization first: There’s no point in DA/DS if your setup isn’t meaningful to begin with. Finally, if you have interpretable subtypes that are in line with your expectation, and these are present across replicates, you should be good to go for DA. 
Sorry for all that- my main point is: reconsider your markers for clustering, and inspect the clusters you have. I have my doubts that with 16 markers and only 1-2 major subtypes, that your current clusters are really meaningful / interpretable. But I could be wrong of course ;)

**Comment by sudu87** [comment]
Great explanation Helena, thanks a lot for investing your time. I will concentrate more on clusters for now.
